### PR TITLE
Clean up Solution Explorer context menus

### LINKED ISSUE #619: Enabling rules from solution explorer doesn't do anything if there's a nowarn in the project
Repro Steps
- Create a project and add the FxCop analyzers to the project. 
- Add some code to fire a violation (say - create an empty enum). Enable CA1008 from solution explorer  and see that the violation is reported.
- Now edit the csproj file and add a <NoWarn>CA1008</NoWarn> to the project.
Note that in Solution explorer, the rule now shows up as disabled. Try to enable it.

**Actual**: Nothing happens.
**Expected**: Some indication of activity. I don't think we want to trawl through the csproj and check to see if there was any configuration - won't be possible if the property was inherited. But I think we should give throw up an error if we try to change the severity and the effective severity is still the same.

### PR DESCRIPTION
**Pull request by tmeschter**
Fixes #619.

I suggest looking at the individual commits one at a time, from top to bottom. Each one is fairly small.

The major work here is to move the context menu items for setting a diagnostic's severity (Error, Warning, etc.) into a sub menu and making it clear that the check marks represent the severity in the rule set specifically, and clicking the menu items will update the rule set file.

Along the way I encountered other bugs I needed to fix to ensure the proper experience here.

### CODE REVIEW

**Comment by tmeschter** [comment]
@srivatsn @jmarolf @shyamnamboodiripad @mavasani @heejaechang @JohnHamby Could you take a look, please?

**Comment by shyamnamboodiripad** [comment]
:+1: except for above comment

**Comment by jmarolf** [comment]
:+1:

File: src/VisualStudio/Core/SolutionExplorerShim/AnalyzersCommandHandler.cs [hunk 1]
```diff
@@ -96,81 +113,199 @@ public void Initialize(IMenuCommandService menuCommandService)
...
+                }
 
-            return menuCommand;
+                return _analyzerFolderContextMenuController;
+            }
+        }
+
+        private bool ShouldShowAnalyzerFolderContextMenu(IEnumerable<object> items)
+        {
+            return items.Count() == 1;
```
Comment (heejaechang) [inline src/VisualStudio/Core/SolutionExplorerShim/AnalyzersCommandHandler.cs]: return !items.Skip(1).Any(); ?

doesn't matter if it is expected to have small number of items though..

File: src/VisualStudio/Core/SolutionExplorerShim/AnalyzersCommandHandler.cs [hunk 2]
```diff
@@ -96,81 +113,199 @@ public void Initialize(IMenuCommandService menuCommandService)
...
+            _addMenuItem.Visible = SelectedProjectSupportsAnalyzers();
+            _addMenuItem.Enabled = _allowProjectSystemOperations;
+        }
+
+        public IContextMenuController AnalyzerContextMenuController
+        {
+            get
             {
-                item.PropertyChanged -= DiagnosticItemPropertyChangedHandler;
+                if (_analyzerContextMenuController == null)
```
Comment (heejaechang) [inline src/VisualStudio/Core/SolutionExplorerShim/AnalyzersCommandHandler.cs]: LazyInitializer ?

File: src/VisualStudio/Core/SolutionExplorerShim/AnalyzersCommandHandler.cs [hunk 3]
```diff
@@ -96,81 +113,199 @@ public void Initialize(IMenuCommandService menuCommandService)
...
-            _setSeverityInfoMenuItem.Checked = AnyDiagnosticsWithSeverity(ReportDiagnostic.Info);
-            _setSeverityHiddenMenuItem.Checked = AnyDiagnosticsWithSeverity(ReportDiagnostic.Hidden);
-            _setSeverityNoneMenuItem.Checked = AnyDiagnosticsWithSeverity(ReportDiagnostic.Suppress);
+            bool errorMenuItemChecked = false;
+            bool warningMenuItemChecked = false;
+            bool infoMenuItemChecked = false;
+            bool hiddenMenuItemChecked = false;
+            bool noneMenuItemChecked = false;
+
+            var workspace = TryGetWorkspace() as VisualStudioWorkspaceImpl;
```
Comment (heejaechang) [inline src/VisualStudio/Core/SolutionExplorerShim/AnalyzersCommandHandler.cs]: how about inverting the ifs ? so it doesnt go nested too deep?

**Comment by heejaechang** [comment]
:+1: my comments are all minor. you can ignore them if you don't feel like changing.

**Comment by tmeschter** [comment]
@shyamnamboodiripad @srivatsn My latest commit updates the check mark behavior. If the build passes, I'm going to merge in the pull request.

**Comment by srivatsn** [comment]
:+1:

**Pull request merged**